### PR TITLE
k3s: document upkeep process

### PR DESCRIPTION
**Pull request by superherointj**
k3s: document upkeep process

The user guidelines has to be improved too. But this is a start.

Let me know of corrections, improvements and objections.

CC K3s maintainers: @euank @yajo @Mic92

### CODE REVIEW

File: pkgs/applications/networking/cluster/k3s/README.md
Comment (yajo) [inline pkgs/applications/networking/cluster/k3s/README.md:83]: I'd clarify this point a little bit more, to avoid having the problem we had in the past (remember https://github.com/NixOS/nixpkgs/issues/213943#issue-1565759962?):

```suggestion
* After every NixOS release, all K3s versions but the latest should be removed by K3s maintainers.
  This way, next NixOS release will have at least one K3s release in common with the previous
  NixOS release. This will provide a smooth upgrade path for users.
```

Comment (superherointj) [reply pkgs/applications/networking/cluster/k3s/README.md:83]: Before a release there might exist multiple K3s versions.
After a release there is going to exist only a single K3s version.
While I understand what you are saying, that is not to be part of 'after a release'.
After a release => Unstable
Unstable starts with a single release of K3s.
Stable will have multiple K3s versions to provide the smooth upgrade path.

Comment (superherointj) [reply pkgs/applications/networking/cluster/k3s/README.md]: @yajo Thanks for your suggestion.

I have tried to integrate what you said. Let me know how it went.

**Review by yajo — APPROVED** [review]

**Comment by superherointj** [comment]
How should `backport` to stable work?
Currently it is broken. But after next release, it will be in sync again. So back-porting might be important.
I don't use stable myself. I continuously pin unstable. (Make unstable stable.) So I don't get to experience the stable issues.

**Comment by superherointj** [comment]
@euank I need your feedback here. Release is happening. And I'm not sure I have correctly documented the process.
It seems I wrongly understood the removal of K3s versions happening after release. And not before release. What should I do?

**Comment by yajo** [comment]
It seems to me like the process you outlined here is very good 🤔

**Comment by euank** [comment]
Thanks for writing this up @superherointj, appreciated!
Apologies for not keeping up with nixpkgs stuff for this release, your help on that is also very much appreciated!

You're right that the intention originally was to remove k3s versions from the next version's release branch, but to leave all supported ones on unstable.

I do feel like deleting all but the newest version on unstable periodically would be really convenient for those of us who only use unstable, but I think we're stuck with following something like I wrote before if we want to allow people to follow k8s's upstream upgrade recommendations, and also stick to nixos stable releases.

I think leaving versions in unstable is also required for backporting patches to older versions while they're supported, i.e. so we can backport 1.28 patches to 23.11 for the month-or-so after 24.05 when it's still supported. I think that's also necessary if we want people to be able to follow upstream k8s's upgrade recommendations (1 version at a time, always using the latest patch version iiuc)

Based on the timeline in https://github.com/NixOS/nixpkgs/issues/303285, it does sound like we have time to sort this out though hopefully!

If we're following what I wrote before, the required next actions would be:

- [ ] Backport k3s 1.28 to release-23.11 (so people can upgrade safely from release-23.11 to 1.29 in release-24.05)
- [ ] As soon as release-24.05 branches off, delete all but k3s 1.29 from it
- [ ] Update the maintainer document a bit more to make this all clearer

Again, apologies for being busy and not putting as much attention into this as I should!

I can open a PR updating some of this documentation in the next couple days as a means to discuss it further, if that sounds good; I'm also happy to change my mind on any of this if we feel like taking some other approach is more pragmatic. The "trimming versions after we branch off" really is a tight timing window, and seems fraught, so I'd love an alternative to that :sweat_smile: 

What do y'all think?
Do the above steps make sense for this release still, or do we have a better option to pursue?

**Comment by superherointj** [comment]
As our time/resources are limited. And there are few people doing upkeep, we should strive to keep it realistic.

I suggest a simple middle way solution:

> To **continuously** remove all and only K3s unstable versions that exist in stable.
> To **try**, but **not require**, to back-port PRs from unstable to stable.
> Whoever is interested in doing upkeep for stable can catch up with back-port of patches and newer versions.

So we don't have to do **special** upkeep for stable. I want to avoid the matrix/doubling of versions in stable, it's more than double the work as patches don't merge nicely.

This would mean, this release would keep `k3s_1_28`, but have all other previous versions removed from unstable.

**Comment by superherointj** [comment]
@euank Thanks for correcting me. As always, glad to have you here!

I've reached a new understanding:

As contributors are limited. To be realistic. The criteria should not rely on individual contributors doing **timely work**.

Basically:
- The pace of removal should match the pace of back-porting.
- After release, all versions from `unstable` that exist in `stable` should be removed.

Currently, for example, would mean:
* `k3s_1_27` would be removed because it exists in `stable`.
* `k3s_1_28` would be kept because it doesn't exist in `stable`. Unless, `k3s_1_28` gets ported to `stable`, then, it should be removed.

This way **special upkeep** for `stable` is optional (desirable, but not mandatory).
From `unstable`, back-porting should be triggered as usual. On failure, the maintainer gets to decide what to do.

Back-porting to stable speeds up package removal. With increased back-porting contributions, version removal will speed up **naturally**. It seems, this criteria **naturally** works for all situations.

I'm against forcing an **external artificial** deadline for packages, while not having the manpower to keep up with such level of contributions. As the package gets used more, new contributors may contribute and then the quality/speed of upkeep will improve naturally. We can't force better output than this. I honestly cannot keep up with much more than what I do already. If some package goes EOL, we have to be realistic. It is just debt that doesn't get solved by a deadline. Polar thinking is bad.

On agreement, I may open a PR fixing documentation with this new understanding.

**Comment by euank** [comment]
I agree reducing maintainership load for k3s is important, and minimizing the number of k3s versions we support is good!

I also think what you've proposed means there's always a supported k3s upgrade path, though it does force people on unstable to upgrade more aggressively.

I think the main reason we ended up where we are in terms of versions was based on discussion in this issue: https://github.com/NixOS/nixpkgs/pull/224483

@yajo, since you had some thoughts then, just to double check, the above proposal works for you right?

I'm okay with the proposal personally, I just use unstable (and can pluck versions off release branches if I need to).
It doesn't feel ideal if we had unlimited manpower, but we don't, so it does seem like it might be a good compromise.

Thanks for writing it up @superherointj, happy to review a docs PR and help push towards that setup if @yajo's also onboard!

**Comment by superherointj** [comment]
>  it does force people on unstable to upgrade more aggressively.

To avoid coupling system upgrade and K3s upgrade, it would be necessary to keep last stable version in unstable. This would avoid breaking K3s instantly on upgrades, and the user could migrate K3s version at his own timing, in a discrete manner. As Kubernetes has differences in application, such as manifests version, CRDs, adds-ons having to catch up to Kubernetes version. This makes sense. So, I agree, that last proposal was too aggressive to unstable (could affect applications not ready for lastest version) and stable (coupled upgrade). That kind of hasty upgrade was in trying to comply with previous nixpkgs release manager demands. Phew...

So the compromise of the compromise, would be to keep at least one/the last version of stable in unstable. Then, stable users can upgrade NixOS without going into a instant K3s upgrade crisis.

Thoughts?

**Comment by euank** [comment]
I'm happy with that proposal, yup!

To be sure I understand the proposal, another way to say it is:

1. We try to keep the latest version of k3s up to date on unstable
2. We try to backport new versions of k3s to the stable release
3. If the newest version on the stable release is different from unstable, we also keep that "latest stable version" on unstable.

This leads to unstable having either 1 or 2 k3s versions, and stable having all versions between the version on unstable when it was cut, up to the current version on unstable, or one version before.

Specifically, that might lead to things like (using hypothetical future versions) "Stable has 1.28 (version when it was cut) + 1.29 + 1.30, unstable has just 1.30, or unstable has 1.30 + 1.31".

Does that sound accurate?

If I'm summarizing what you're saying correctly, that sounds good to me!

**Comment by superherointj** [comment]
First, I'm sorry for being repetitive and stating obvious things. Communication is hard. I'm repeating some understandings for emphasis and checking that our understanding is shared. I hope it doesn't get too annoying, but I couldn't do much better for the moment.

>     1. We try to keep the latest version of k3s up to date on unstable

Yes.

>     2. We try to backport new versions of k3s to the stable release

Yes, `stable` will receive back-ports whenever possible. (All that merge fine for sure.)
Manual work for back-porting would be optional and executed by interested contributors.

>     3. If the newest version on the stable release is different from unstable, we also keep that "latest stable version" on unstable.

Yes, `unstable` would keep K3s last `latest-stable` version.

> This leads to unstable having either 1 or 2 k3s versions, 

Yes. Whenever `latest-stable` version would equate to unstable version, `unstable` would have a single K3s version.

> and stable having all versions between the version on unstable when it was cut,

Yes.

> up to the current version on unstable, or one version before.

Back-porting enables version removal from `unstable` at next release.
A release (without extra backports) automatically defines the **slowest** version removal pace.
Optionally, any newer version could be back-ported to `stable`, if there are interested contributors. But it's not guaranteed. So, the stable set of versions may lag in reference to `unstable`. This is not a problem.
The **reference frame** I have been using to think is `unstable` (and not `stable`, as you did):

As all changes are done to `unstable`, `stable` is a past mirror image of `unstable`.
The K3s package removal happens only at `unstable`. And `stable` inherits that from `unstable` at release.

So, `unstable` only needs to keep the `latest-stable` version for system upgrades happening smoothly.
The many versions `stable` has is not important for the upgrade process. At `unstable`, we look backwards to `stable` to check the latest version `stable` supports, and we keep just that.

> Specifically, that might lead to things like (using hypothetical future versions) "Stable has 1.28 (version when it was cut) + 1.29 + 1.30, unstable has just 1.30, or unstable has 1.30 + 1.31".

I was thinking from `unstable` reference and you were thinking `stable` reference, a reversed frame of reference.

- Unstable will always have `latest-stable` version.
- Stable will have the versions from when `unstable release` happened + (optional) backports.

Considering `latest-unstable` as `1.31`, looking back to `stable` to check:
- If `latest-stable` is 1.31, `unstable` keeps 1.31 only.
- If `latest-stable` is 1.30, `unstable` keeps 1.30+1.31.
- If `latest-stable` is 1.29, `unstable` keeps 1.29+1.30+1.31.
- If `latest-stable` is 1.28, `unstable` keeps 1.28+1.29+1.30+1.31.

So, yeah, the way reduce versions in `unstable` is to do back-porting. Then, it's desirable to do back-porting to limit the versions in unstable. On passive upkeep, the automatic release pace of removal happens.

> Does that sound accurate?

I think, we are in sync. If not, we'll be now.

> If I'm summarizing what you're saying correctly, that sounds good to me!

I'll prepare the documentation. Then, we can iterate in the PR.

Other considerations:
- We should recommend users to continuously upgrade K3s. (To be at last available K3s version.)

**Comment by yajo** [comment]
It's becoming a bit complex to understand for me, I guess It'll be easier on the PR.

However, to put it simple, I a gree with this sentence: 

> To avoid coupling system upgrade and K3s upgrade, it would be necessary to keep last stable version in unstable

Reasoning:

- A system upgrade is hard enough.
- If something breaks, it'll be easier to know if it's k3s itself or the rest of your NixOS.

So, unstable and stable should share at least one minor k3s version.

**Comment by superherointj** [comment]
> However, to put it simple, **unstable and stable should share at least one minor k3s version**.

Yes. Let's sum up with that.

**Pull request merged**
